atd-etl/cris_import/parse_cris_schema.py

```diff
@@ -0,0 +1,688 @@
+    for worksheet in workbook.worksheets:
+        print("")
+        print("Title: ", worksheet.title.lower())
+        match = re.search(r"(\w+)_LKP", worksheet.title)
+        lookup_table = match.group(1).lower() if match else None
+        if lookup_table:
+            if only_process_these_tables and not lookup_table in only_process_these_tables:
+                continue 
+
+            drop = f"drop table if exists cris_lookup.{lookup_table} cascade;"
+            drop_cursor = pg.cursor()
+            print(f"Drop: {drop}")
+            drop_cursor.execute(drop)
+            drop_cursor.close()
+            pg.commit()
+
+            drop = f"drop table if exists vz_lookup.{lookup_table} cascade;"
+            drop_cursor = pg.cursor()
+            print(f"Drop: {drop}")
+            drop_cursor.execute(drop)
+            drop_cursor.close()
+            pg.commit()
+
+            drop_sequence = f"DROP SEQUENCE IF EXISTS lookup.{lookup_table}_global_id"
+            drop_sequence_cursor = pg.cursor()
+            print(f"Drop: {drop_sequence}")
+            drop_sequence_cursor.execute(drop_sequence)
+            drop_sequence_cursor.close()
+            pg.commit()
+
+            create_sequence = f"CREATE SEQUENCE lookup.{lookup_table}_global_id"
+            create_sequence_cursor = pg.cursor()
+            print(f"Create Sequence: {create_sequence}")
+            create_sequence_cursor.execute(create_sequence)
+            create_sequence_cursor.close()
+            pg.commit()
+
+            if lookup_table == "state":
+                populate_state_table(worksheet, lookup_table, pg)
+            elif lookup_table == "veh_mod_year":
+                populate_veh_mod_year_table(worksheet, lookup_table, pg)
+            elif lookup_table == "cntl_sect":
+                populate_cntl_sect_table(worksheet, lookup_table, pg)
+            else:
+                populate_table(worksheet, lookup_table, pg)
+
+def check_cris_lookup_values_unicity(file_path, pg):
+    workbook = load_workbook(filename=file_path)
+    for worksheet in workbook.worksheets:
+        print("")
+        print("Title: ", worksheet.title.lower())
+        match = re.search(r"(\w+)_LKP", worksheet.title)
+        lookup_table = match.group(1).lower() if match else None
+        if lookup_table == 'cntl_sect':
+            continue
+        if lookup_table:
+            print(lookup_table)
+            sql = f"""select count(id)
+                from lookup.{lookup_table}
+                group by cris_id, source, vz_id
+                having count(id) > 1;
+            """
+            cursor = pg.cursor(cursor_factory=psycopg2.extras.DictCursor)
+            cursor.execute(sql)
+            rows = cursor.fetchall()
+            cursor.close()
+            if len(rows) > 0:
+                print(rows)
+                print("ERROR: Duplicate values in lookup table")
+                exit(1)
+
+def create_materialized_views(file_path, pg):
```

Comment:
I inspected the DB and see the the `lookup` schema has these materialized view, and the `cris_lookup` and `vz_lookup` have tables. What are the relationships between these three schema? I'm also curious about the materialized views out of curiosity. I just don't have a good enough feel about postgres to connect the dots.

Reply:
Sure thing, here's the three schema and their relationships:

### `cris_lookup`
This is a schema holding the CRIS provided lookup tables and their values. They are populated from parsing a slightly-modified version of a XLSX file that CRIS makes available [here](https://ftp.txdot.gov/pub/txdot-info/trf/crash_statistics/automated/standardextractfilespecification.xlsx). 

In the future, I think it'd be neat to be checking and possibly maintaining the contents of these tables by examining the `lookup` CSVs provided in our daily exports. 

These tables match up, one-to-one, to a table in the `vz_lookup` schema, and entries in both tables in each pair, share a common ID sequence so they can define a always-unique column between each other.

Most of the tables have a similar shape, but there are some outliers, and you can see them being handled with special handler functions in this program.

### `vz_lookup`
This is a schema holding the VZ provided lookup table values. By and large, the shape of these tables is predictable based on the columns that CRIS gave us. Also, there's nothing preventing us from adding VZ specific associated data which is a feature of the LDM. 

These tables are not populated by this script, because it's interpreting CRIS provided data. However, we will need to add some specific VZ provided values that have been added into our production database. 

When we search for orphaned FK values before we launch🎗️, it'll reveal which ones VZ has requested and used.

### `lookup` 
* This contains materialized views that unify the `vz_lookup` and `cris_lookup` pairs of tables. The query to `create` each materialized view is constructed programmatically, so they may be easier to review via a DB client and inspecting the materialized view.
* Materialized views are used here because lookup tables change their values very slowly, and it's well worth having to `refresh materialized view ...` when we make the very occasional change and pick up the ability to index these tables.
* This feels like it should get a better name, but I'm not sure ... `ldm_lookup`? It feels redundant and not like adding more information by adding the `ldm_`, but I'm not sure.

Reply:
awesome, never commented here that this helped me explore the DB a lot after running through the LDM transformation steps. We should copy/paste your explanation here into some docs!